docs/RFCS/raft_consistency_checker.md

```diff
@@ -0,0 +1,86 @@
+- Feature Name: Raft consistency checker
+- Status: draft
+- Start Date: 2016-02-10
+- Authors: Ben Darnell, Bram Grunier, Vivek Menezes
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #837
+
+# Summary
+
+An online system that checks that periodically checks that range replicas
+are consistent; a snapshot for a range at a specific point in the raft log
+ has the same sha on all replicas.
+
+# Motivation
+
+Consistency! Correctness at scale.
+
+# Detailed design
+
+The proposal is for each range replica leader to periodically run the
+consistency checker. The consistency checker runs in three phases:
+
+1. All replicas agree on the snapshot at which a sha is to be computed.
+2. A sha is computed on all replicas.
+3. The leader shares its sha with all replicas; a replica computing a
+different sha from the one supplied by the leader alerts and eutanizes
+the node.
+
+The above is implemented by introducing two raft commands:
+
+1. CHECK_CONSISTENCY_BEGIN (BEGIN): A replica on receiving this command
+immediately acknowledges it has executed it, and asynchronously computes the sha
+for the current snapshot. It continues executing commands while it computes the
+sha on the side. Once the sha is computed it compares the BEGIN command id along
+with the computed sha to the one received from the leader in the second command
+below.
+2. CHECK_CONSISTENCY_CHECKSUM (CHECKSUM): This command contains the id for the
+BEGIN command it is referring to, and the sha that it has computed at the BEGIN
+command snapshot on the leader. On receiving this command the replica saves the
+command id and sha supplied, and compares the sha against the one it eventually
+computes. Under the rare circumstances that it sees a difference, it will alert
+and euthanize the node.
+
+Each range replica has the ability to keep track of multiple sha computations
+that are being computed or recently computed. The shas get GC-ed after about
+30 minutes.
+
+The periodic consistency checker is run within a scanner that scans through all
+the local replicas and runs the consistency checker for all leader replicas.
+The scanner runs in a continuous loop with equal time spacing between
+replicas such that it spans the entire periodicity interval.
+
+Noteworthy scenarios:
+
+1. The leader dies between BEGIN and CHECKSUM: The replicas, including the new
+leader, will continue computing the sha. The new leader will send a CHECKSUM
```

Comment:
why not just let things be until the next queue run on the new leader begins with BEGIN again? That seems simpler and about as effective as introducing state and hence complexity.